Archive/Imo/Imo2025Q1.lean

```diff
@@ -0,0 +1,1309 @@
+    obtain ha | hb := h
+    · use (w₁ 1 - w₂ 1) / a
+      · rw [hw₁₂]
+        nth_rw 2 [vec_repr w₁, vec_repr w₂]
+        rw [vec_mul, vec_sub, vec_eq]
+        constructor <;> (field_simp; try linarith)
+    · use (w₂ 0 - w₁ 0) / b
+      · rw [hw₁₂]
+        nth_rw 2 [vec_repr w₁, vec_repr w₂]
+        rw [vec_mul, vec_sub, vec_eq]
+        constructor <;> (field_simp; try linarith)
+  · intro ⟨k, hkw⟩
+    rw [AffineSubspace.mem_direction_iff_eq_vsub hv_nonempty]
+    simp only [Set.Nonempty, SetLike.mem_coe] at hv_nonempty
+    obtain ⟨v₁, hv₁⟩ := hv_nonempty
+    use v₁
+    refine ⟨hv₁, v₁ - w, ?_, by simp⟩
+    rw [vec_repr v₁, point_on_line] at hv₁
+    rw [vec_repr v₁, ← hkw, vec_mul, vec_sub, point_on_line]
+    ring_nf; exact hv₁
+
+/-- The rank of the line `a * x + b * y + c = 0` is `1` when `a` and `b` are not both zero -/
+lemma line_rank (a b c : ℝ) (h : a ≠ 0 ∨ b ≠ 0) : finrank ℝ (line a b c).direction = 1 := by
+  have hv_mem : !₂[-b, a] ∈ AffineSubspace.direction (line a b c) := by
+    rw [line_direction (h := h)]
+    use 1; simp
+  rw [finrank_eq_one_iff']
+  refine ⟨⟨!₂[-b, a], hv_mem⟩, by simp; tauto, ?_⟩
+  simp only [SetLike.mk_smul_mk, Subtype.forall, Subtype.mk.injEq]
+  exact fun w ↦ (line_direction a b c h w).mp
+
+/-- The coeffecients of a line a * x + b * y + c = 0 -/
+structure LineCoeffs where
+  a : ℝ
+  b : ℝ
+  c : ℝ
+
+/-- Build a line from a `LineCoeffs` -/
+noncomputable def line' (coeff : LineCoeffs) := line coeff.a coeff.b coeff.c
+
+def edgeCoeffs (n : ℕ) (d : Fin 3) : LineCoeffs := match d with
+| 0 => ⟨1, 0, -1⟩
+| 1 => ⟨0, 1, -1⟩
+| 2 => ⟨1, 1, -(n + 1)⟩
+
+/-- The three lines on the edges of the triangular integer grid, each side having n points:
+`x - 1 = 0`; `y - 1 = 0`; `x + y - (n + 1) = 0`.
+These lines are named `edgeLine n 0`, `edgeLine n 1`, `edgeLine n 2`, resp. -/
+noncomputable def edgeLine (n : ℕ) (d : Fin 3) := line' (edgeCoeffs n d)
+
+/-- The y axis is the line `x = 0`. -/
+lemma y_ax_line : line 1 0 0 = yAxis := by simp [line, yAxis]
+
+/-- The x axis is the line `y = 0` -/
+lemma x_ax_line : line 0 1 0 = xAxis := by simp [line, xAxis]
+
+/-- The line `x + y = 0` -/
+lemma xy0_line : line 1 1 0 = linexy0 := by simp [line, linexy0]
+
+/-- Preparation lemma for checking if two lines are parallel -/
+lemma line_para' (a b a' b' : ℝ) (h' : a' ≠ 0 ∨ b' ≠ 0)
+    (hab : a * b' = a' * b) (w : Plane) :
+    (∃ (k : ℝ), k • !₂[-b, a] = w) → (∃ (k : ℝ), k • !₂[-b', a'] = w) := by
+  intro ⟨k, hk₁⟩
+  rw [vec_mul, vec_repr w, vec_eq] at hk₁
+  obtain ha | hb := h'
+  · use k * a / a'
+    rw [vec_mul, vec_repr w, vec_eq]
+    grind
+  · use k * b / b'
+    rw [vec_mul, vec_repr w, vec_eq]
+    grind
+
+/-- Two non-degenerate lines `a * x + b * y + c = 0` and `a' * x + b' * y + c = 0`
+are parallel iff `a * b' = a' * b`. -/
+lemma line_para (a b c a' b' c' : ℝ) (h : a ≠ 0 ∨ b ≠ 0) (h' : a' ≠ 0 ∨ b' ≠ 0) :
+    line a b c ∥ line a' b' c' ↔ a * b' = a' * b := by
```

Comment:
I don't think generic material about plane geometry like this should be developed in the `Archive/` directory. It should be PR'd to Mathlib first, and then the IMO solution can be contributed when the library is sufficiently developed.